modules/consensus/synchronize.go

```diff
@@ -11,54 +11,76 @@ import (
 )
 
 const (
-	MaxCatchUpBlocks       = 10
-	MaxSynchronizeAttempts = 8
+	// MaxCatchUpBlocks is the maxiumum number of blocks that can be given to
+	// the consensus set in a single iteration during the initial blockchain
+	// download.
+	MaxCatchUpBlocks = 10
 )
 
-// blockHistory returns up to 32 BlockIDs, starting with the 12 most recent
-// BlockIDs and then doubling in step size until the genesis block is reached.
-// The genesis block is always included. This array of BlockIDs is used to
-// establish a shared commonality between peers during synchronization.
-func (s *ConsensusSet) blockHistory() (blockIDs [32]types.BlockID) {
-	knownBlocks := make([]types.BlockID, 0, 32)
+// blockHistory returns up to 32 block ids, starting with recent blocks and
+// then proving exponentially increasingly less recent blocks. The genesis
+// block is always included as the last block. This block history can be used
+// to find a common parent that is reasonably recent, usually the most recent
+// common parent is found, but always a common parent within a factor of 2 is
+// found.
+func blockHistory(tx *bolt.Tx) (blockIDs [32]types.BlockID) {
+	height := blockHeight(tx)
 	step := types.BlockHeight(1)
-	for height := s.height(); ; height -= step {
-		// after 12, start doubling
-		knownBlocks = append(knownBlocks, s.db.getPath(height))
-		if len(knownBlocks) >= 12 {
+	// The final step is to include the genesis block, which is why `step <
+	// height` is used as opposed to `step <= height`.
+	for i := 0; i < 31; i++ {
+		// Include the next block.
+		blockID, err := getPath(tx, height)
+		if build.DEBUG && err != nil {
+			panic(err)
+		}
+		blockIDs[i] = blockID
+
+		// Determine the height of the next block to include and then increase
+		// the step size. The height must be decreased first to prevent
+		// underflow.
+		//
+		// `i >= 9` means that the first 10 blocks will be included, and then
+		// skipping will start.
+		if i >= 9 {
```

Comment:
Old code did 13 blocks, followed by 19 exponential blocks, one of which was the genesis block. 2^18 is only 250k, which is not enough. Need to get into the millions.

9 is 10 blocks, genesis makes 11, 2^21 = 2 million, which is more acceptable.